### PR TITLE
fix: namespace-scoped mode by service_controller

### LINKED ISSUE #433: Quickstart broken: `modelmesh-serving` svc not created in `--namespace-scope-mode`, ISVCs never get to `READY` state
**Describe the bug**

The [quickstart install](https://github.com/kserve/modelmesh-serving/blob/main/docs/quickstart.md#run-the-installation-script) instructions no longer work correctly. After deploying a model, the `InferenceService` does not get into `READY` state. Inference requests cannot be performed. 

**To Reproduce**

1. Follow the Quickstart guide the point of [deploying a model](https://github.com/kserve/modelmesh-serving/blob/main/docs/quickstart.md#2-deploy-a-model) and wait for the runtime pods to be `Running`.

```shell
$ kubectl get pods

NAME                                              READY   STATUS    RESTARTS   AGE
etcd-558fb445c-ht596                              1/1     Running   0          9h
minio-774d48cf7-nbkd7                             1/1     Running   0          9h
modelmesh-controller-86dd54cc54-xhnvv             1/1     Running   0          9h
modelmesh-serving-mlserver-1.x-647d4f8f9c-h5b49   4/4     Running   0          26m
modelmesh-serving-mlserver-1.x-647d4f8f9c-hv6ht   4/4     Running   0          26m
```

2. Wait a few minutes (or hours) and check the status of the newly created `InferenceService`.

```shell
$ kubectl get isvc

NAME                   URL   READY   PREV   LATEST   PREVROLLEDOUTREVISION   LATESTREADYREVISION   AGE
example-sklearn-isvc                                                                               26m
```

3. Describe the ISVC, find no events:

```shell
$ kubectl describe isvc example-sklearn-isvc

Name:         example-sklearn-isvc
Namespace:    modelmesh-serving
Labels:       <none>
Annotations:  serving.kserve.io/deploymentMode: ModelMesh
              serving.kserve.io/secretKey: localMinIO
API Version:  serving.kserve.io/v1beta1
Kind:         InferenceService
Metadata:
  Creation Timestamp:  2023-09-20T03:21:13Z
  Generation:          2
  Managed Fields:
    API Version:  serving.kserve.io/v1beta1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
          f:serving.kserve.io/deploymentMode:
          f:serving.kserve.io/secretKey:
      f:spec:
        .:
        f:predictor:
          .:
          f:model:
            .:
            f:modelFormat:
              .:
              f:name:
            f:storageUri:
    Manager:         kubectl-client-side-apply
    Operation:       Update
    Time:            2023-09-20T03:46:54Z
  Resource Version:  61292
  UID:               2d2acd78-e0ce-4864-bd3f-54944d02ae85
Spec:
  Predictor:
    Model:
      Model Format:
        Name:       sklearn
      Storage Uri:  s3://modelmesh-example-models/sklearn/mnist-svm.joblib
Events:             <none>


```

4. Ignore all that for a minute and go on to create the port-forward:

```shell
$ kubectl port-forward --address 0.0.0.0 service/modelmesh-serving 8033 -n modelmesh-serving

Error from server (NotFound): services "modelmesh-serving" not found

```

5. Check the service list:

```shell

$ kubectl get svc

NAME                               TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)    AGE
etcd                               ClusterIP   172.21.141.59    <none>        2379/TCP   9h
minio                              ClusterIP   172.21.11.102    <none>        9000/TCP   9h
modelmesh-webhook-server-service   ClusterIP   172.21.232.100   <none>        9443/TCP   9h

```

- **Note**, the service "modelmesh-serving" does not exist.

**Expected behavior**

Up until release `v0.11.0` when following the instructions in the Quickstart guide the inference service was created successfully, port-forward worked and inference requests returned expected response.

**Additional context**

How do the FVT test cases work? Curiously, when using the FVT install instructions  using the `--fvt` flag (cluster-scoped install), the `modelmesh-serving` service does get created and the ISVC transitions to `READY` state okay.

```shell
$ kubectl get svc

NAME                               TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)                      AGE
etcd                               ClusterIP   172.21.33.247    <none>        2379/TCP                     2m7s
minio                              ClusterIP   172.21.215.204   <none>        9000/TCP                     2m6s
modelmesh-serving                  ClusterIP   None             <none>        8033/TCP,8008/TCP,2112/TCP   75s
modelmesh-webhook-server-service   ClusterIP   172.21.82.140    <none>        9443/TCP                     104s


```

**Environment (please complete the following information):**

- OS: macOS client side, Ubuntu 20 server side
- Browser: [n/a]
- Version: MM latest of `main`, Kubernetes 1.25, 1.26, 1.27

---

**UPDATE 2023-09-21:**

FYI @Jooho

I went through some of the recent changes and it looks like PR #397 introduced this changed behavior. Could it be that the way your logic changes on how namespaces are reconciled affects how/when/whether the `modelmesh-serving` service gets created?

```
W0921   1 reflector.go:424] pkg/mod/k8s.io/client-go@v0.26.4/tools/cache/reflector.go:169: 
  failed to list *v1.Namespace: namespaces is forbidden: 
    User "system:serviceaccount:modelmesh-serving:modelmesh-controller" cannot list 
      resource "namespaces" in API group "" at the cluster scope

E0921   1 reflector.go:140] pkg/mod/k8s.io/client-go@v0.26.4/tools/cache/reflector.go:169:
  Failed to watch *v1.Namespace: failed to list *v1.Namespace: namespaces is forbidden:
    User "system:serviceaccount:modelmesh-serving:modelmesh-controller" cannot list
      resource "namespaces" in API group "" at the cluster scope
```

### PR DESCRIPTION
**Pull request by Jooho**
Add a ClusterScope condition statement to list namespace that is only allowed for cluster scope.

Resolves #433

### CODE REVIEW

**Comment by Jooho** [comment]
@ckadner fyi

**Comment by Jooho** [comment]
@ckadner FVT test passed on local minikube. I think it should work with gitaction eventually. could you  please trigger the fvt test? I think I can not retrigger the gitaction.

File: controllers/service_controller.go [hunk 1]
```diff
@@ -247,10 +247,17 @@ func (r *ServiceReconciler) reconcileService(ctx context.Context, mms *mmesh.MMS
...
 	//  - when the namespace is under a Terminating state.
 	n := &corev1.Namespace{}
-	if err := r.Client.Get(ctx, types.NamespacedName{Name: namespace}, n); err != nil {
-		return nil, err, false
+	if r.ClusterScope {
+		if err := r.Client.Get(ctx, types.NamespacedName{Name: namespace}, n); err != nil {
+			return nil, err, false
+		}
+	} else {
+		n = &corev1.Namespace{
```
Comment (ckadner) [inline controllers/service_controller.go:255]: why reassign `n` here? Wouldn't the Client.Get call just above have returned the namespace object with the name?

File: controllers/service_controller.go [hunk 2]
```diff
@@ -247,10 +247,17 @@ func (r *ServiceReconciler) reconcileService(ctx context.Context, mms *mmesh.MMS
...
+			ObjectMeta: metav1.ObjectMeta{
+				Name: namespace,
+			},
+		}
 	}
-
 	if !modelMeshEnabled(n, r.ControllerDeployment.Namespace) {
```
Comment (ckadner) [inline controllers/service_controller.go:261]: in namespace scope, `n` would be empty here, no?

**Review by ckadner — COMMENTED** [review]
Thanks @Jooho for providing a fix quickly! 

A few questions:

File: controllers/service_controller.go (hunk 1 above)
Comment (Jooho) [reply controllers/service_controller.go]: @ckadner Thanks for your review.

I have thoroughly reviewed the logic for cluster scope and namespace scope. As a result, I have made changes to ensure that this logic is executed only when in cluster scope, as in the case of namespace scope, the namespace is always the same as the controller namespace, so there is no need to retrieve the namespace object separately.

Additionally, I identified an issue where if a namespace with the "modelmesh-enabled=true" label is suddenly deleted, the code was still attempting to retrieve the namespace, causing errors in the serving runtime controller([code](https://github.com/kserve/modelmesh-serving/blob/3f018571f0dd79bb9905dde6855642c73478c048/controllers/servingruntime_controller.go#L95)). To address this, I updated [the modelmeshEnabled function](https://github.com/kserve/modelmesh-serving/pull/435/files#diff-f5aad6c746366d88daae1f63f8dcff26e2205cb9eb6b455bee29d67f0edc6573R46) in the utils.go file to consider that if the namespace does not exist, it should be treated as if modelmesh was not enabled in that namespace.

I conducted end-to-end testing (FVT) for both namespace scope and cluster scope modes locally, and both of them passed successfully.

please review it again @ckadner @tjohnson31415

Comment (ckadner) [reply controllers/service_controller.go]: looks good now. Thanks Jooho

**Review by ckadner — APPROVED** [review]
Thanks @Jooho -- this looks good now.

/lgtm

**Pull request merged**